File: H-index-2.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public int hIndex(int[] citations) {
        if(citations == null || citations.length == 0){
            return 0;
        }
        int l = 0;
        int r = citations.length-1 ;
        while(l<=r){
            int m = l + (r - l) /2; // prevnt integer overflow
            if(citations[m] >= citations.length - m){
                r = m - 1;
            } 
            else{
                l = m + 1; 
            }
        }
        return citations.length - l;
    }
}

File: H Index 2.cs
/*
  Time complexity: O(log n)
  Space complexity: O(1)

*/

public class Solution {
    public int HIndex(int[] citations) {
        int low = 0;
        int n = citations.Length;
        int high = n-1;
        int mid = 0;

        while(low<=high)
        {
            mid = low+(high-low)/2;
            int diff = n - mid;
            if(diff==citations[mid])
                return diff;
            else
                if(diff>citations[mid])
                    low=mid+1;
                else    
                    high = mid-1;
        }
        return n-low;
    }
}
